src/main/java/org/utplsql/api/db/DynamicParameterList.java

```diff
@@ -0,0 +1,91 @@
+package org.utplsql.api.db;
+
+import oracle.jdbc.OracleConnection;
+
+import java.sql.CallableStatement;
+import java.sql.SQLException;
+import java.sql.Types;
+import java.util.LinkedHashMap;
+import java.util.stream.Collectors;
+
+public class DynamicParameterList {
+
+    private LinkedHashMap<String, DynamicParameter> params;
+
+    interface DynamicParameter {
+        void setParam( CallableStatement statement, int index ) throws SQLException;
+    }
+
+    static class DynamicStringParameter implements DynamicParameter {
+        private final String value;
+
+        DynamicStringParameter( String value ) {
+            this.value = value;
+        }
+
+        @Override
+        public void setParam(CallableStatement statement, int index) throws SQLException {
+            if ( value == null ) {
+                statement.setNull(index, Types.VARCHAR);
+            } else {
+                statement.setString(index, value);
+            }
+        }
+    }
+    static class DynamicIntegerParameter implements DynamicParameter {
+        private final Integer value;
+
+        DynamicIntegerParameter( Integer value ) {
+            this.value = value;
+        }
+
+        @Override
+        public void setParam(CallableStatement statement, int index) throws SQLException {
+            if ( value == null ) {
+                statement.setNull(index, Types.INTEGER);
+            } else {
+                statement.setInt(index, value);
+            }
+        }
+    }
+    static class DynamicArrayParameter implements DynamicParameter {
+        private final Object[] value;
+        private final String customTypeName;
+        private final OracleConnection oraConnection;
+
+        DynamicArrayParameter( Object[] value, String customTypeName, OracleConnection oraConnection ) {
+            this.value = value;
+            this.customTypeName = customTypeName;
+            this.oraConnection = oraConnection;
+        }
+
+        @Override
+        public void setParam(CallableStatement statement, int index) throws SQLException {
+            if ( value == null ) {
+                statement.setNull(index, Types.ARRAY, customTypeName);
+            } else {
+                statement.setArray(
+                        index, oraConnection.createOracleArray(customTypeName, value)
+                );
+            }
+        }
+    }
+
+    DynamicParameterList(LinkedHashMap<String, DynamicParameter> params) {
```

Comment:
if the builder is the only way we allow to create such objects, then Builder class should be moved inside this class as static nested class and the constructor should be private.

Reply:
Yep, absolutely, thanks!